Check for __GNUC__ to use GNU C atomic buildins

levitte opened the pull request:
Note: we trust any other compiler that fully implements GNU extension
to define **GNUC**

RT#4642

ghedo commented on crypto/threads_pthread.c:
```diff
@@ -11,6 +11,10 @@
 #if defined(OPENSSL_THREADS) && !defined(CRYPTO_TDEBUG) && !defined(OPENSSL_SYS_WINDOWS)
 
+# if __STDC_VERSION__ >= 201112L && !defined(__STDC_NO_ATOMICS__)
+#  include <stdatomics.h>
```
I think this is supposed to be `stdatomic.h`.

levitte replied:
You're entirely correct.  I get confused by `__STDC_NO_ATOMICS__` (note ending S).  Fix coming up.

levitte committed:
CRYPTO_atomic_add(): check that the object is lock free

If not, fall back to our own code, using the given mutex

levitte committed:
CRYPTO_atomic_add(): use acquire release memory order rather than relaxed

For increments, the relaxed model is fine.  For decrements, it's
recommended to use the acquire release model.  We therefore go for the
latter.

levitte committed:
Check for __GNUC__ to use GNU C atomic buildins

Note: we trust any other compiler that fully implements GNU extension
to define __GNUC__

RT#4642

levitte commented:
After much internal deliberation and noting that the C11 compliant changes needed a larger change, we've ended up just doing better checks for the GNU atomic implementation.

levitte commented:
Has now been merged into master.  Closing.
